crates/flox-rust-sdk/src/models/environment.rs

```diff
@@ -33,75 +35,112 @@ pub enum InstalledPackage {
+        packages: impl IntoIterator<Item = FloxPackage>,
+        nix: &NixCommandLine,
+        system: impl AsRef<str> + Send,
+    ) -> Result<&mut Self, EnvironmentError2>;
+
+    /// Uninstall remove pacakges from the environment atomically
+    async fn uninstall(
+        &mut self,
+        packages: impl IntoIterator<Item = FloxPackage>,
+        nix: &NixCommandLine,
+        system: impl AsRef<str> + Send,
+    ) -> Result<&mut Self, EnvironmentError2>;
+
+    /// Return the [EnvironmentRef] for the environment for identification
+    fn environment_ref(&self) -> &EnvironmentRef;
+
+    /// Read the catalog for this environment
+    ///
+    /// The catalog contains information about the locked sources of installed packages.
+    async fn catalog(
+        &self,
+        nix: &NixCommandLine,
+        system: impl AsRef<str>,
+    ) -> Result<EnvCatalog, EnvironmentError2>;
+
+    /// List the installed packages
+    async fn packages(
+        &self,
+        nix: &NixCommandLine,
+        system: impl AsRef<str>,
+    ) -> Result<Vec<FloxPackage>, EnvironmentError2>;
+
+    /// Create a temporary environment that can be modified freely
+    async fn modify_in(
+        &mut self,
+        path: impl AsRef<Path>,
+    ) -> Result<Self::ConcreteTemporary, EnvironmentError2>;
+
+    /// Apply the changes made in a temporary environment
+    fn replace_with(
+        &mut self,
+        temporary_environment: Self::ConcreteTemporary,
+    ) -> Result<(), EnvironmentError2>;
+
+    /// Delete the Environment
+    fn delete(self) -> Result<(), EnvironmentError2>;
+}
+
+#[async_trait]
+pub trait TemporaryEnvironment {
+    async fn set_environment(
+        &mut self,
+        mut flox_nix_content: impl std::io::Read + Send,
+        nix: &NixCommandLine,
+        system: impl AsRef<str> + Send,
+    ) -> Result<(), EnvironmentError2>;
+}
+
+/// Struct representing a local environment in a given location
 #[derive(Debug)]
-pub struct Environment<S> {
+pub struct PathEnvironment<S> {
     /// absolute path to the environment, typically within `<...>/.flox/name`
     path: PathBuf,
-    /// Access state of the environment
-    ///
-    /// Implementations distinguish whether whe can [Modify] or only [Read] an environment
+    /// The [EnvironmentRef] this env is created from (and validated against)
+    environment_ref: EnvironmentRef,
     state: S,
 }
 
-impl<S> PartialEq for Environment<S> {
+/// Changes to environments should be atomic,
+/// so that possibly breaking modifications to the environment can be safely discarded.
+/// When environment objects are created with [EnvironmentState] [Original] they
+/// are in a "reading" state.
+/// To make modifications we copy the environment into a temporary sandbox.
+/// Within the sandbox we can make modifications and verify them by building the environment.
+/// When verified, we can move the sandboxed environment back to its original location.
+#[derive(Debug)]
+pub struct Original {
```

Comment:
I think the temporary directory should live on whatever struct is implementing `Environment`. Using typestates is already using a bit of our fanciness budget, so I'd like to keep the state structs as minimal as possible. In other words, I think it's fine to require that whatever is implementing `Environment` to know how to turn itself into a temporary environment.

If you remember what @tomberek said about wanting to be able to test that changes are made in-place, whatever test environment is implementing `Environment` could just have a no-op for `modify_in` or `make_temporary` so that `install` etc operate on the existing directory.